include/nbl/builtin/hlsl/workgroup2/shared_scan.hlsl

```diff
@@ -0,0 +1,466 @@
+// Copyright (C) 2025 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+#ifndef _NBL_BUILTIN_HLSL_WORKGROUP2_SHARED_SCAN_INCLUDED_
+#define _NBL_BUILTIN_HLSL_WORKGROUP2_SHARED_SCAN_INCLUDED_
+
+#include "nbl/builtin/hlsl/cpp_compat.hlsl"
+#include "nbl/builtin/hlsl/workgroup/broadcast.hlsl"
+#include "nbl/builtin/hlsl/glsl_compat/subgroup_basic.hlsl"
+#include "nbl/builtin/hlsl/subgroup/ballot.hlsl"
+#include "nbl/builtin/hlsl/subgroup2/arithmetic_portability.hlsl"
+#include "nbl/builtin/hlsl/mpl.hlsl"
+
+namespace nbl 
+{
+namespace hlsl
+{
+namespace workgroup2
+{
+
+namespace impl
+{
+template<uint16_t WorkgroupSizeLog2, uint16_t SubgroupSizeLog2>
+struct virtual_wg_size_log2
+{
+    NBL_CONSTEXPR_STATIC_INLINE uint16_t levels = conditional_value<(WorkgroupSizeLog2>SubgroupSizeLog2),uint16_t,conditional_value<(WorkgroupSizeLog2>SubgroupSizeLog2*2+2),uint16_t,3,2>::value,1>::value;
+    NBL_CONSTEXPR_STATIC_INLINE uint16_t value = mpl::max_v<uint32_t, WorkgroupSizeLog2-SubgroupSizeLog2, SubgroupSizeLog2>+SubgroupSizeLog2;
```

Comment:
you need a static assert here that `WorkgroupSizeLog2>=_SubgroupSizeLog2`

Reply:
the computation of `value` is wrong, see the spreadsheet https://docs.google.com/spreadsheets/d/1tPbXd8AgSl3U0XSszx5TFiugP_ls_nBlWPqvvctSBnw/edit?usp=sharing

I marked in red the virtual workgroup counts which get needlessly inflated, e.g.:
- WG size 128 with SG size 128 will spit out a VG of 16k instead of 1, even though the level count is 1
- 64, 64 does 4096
- 32,32 does 1024
and so on

I'd probably explicit specialize this struct for the cases needing 1 level and 3 level scans.

Reply:
note that is always seems to happen when subgrouplog2==workgrouplog2 and level=1

Reply:
I think the virtual workgroup size should be
```
SubgroupLog2*2 if WorkgroupLog2 <   SubgroupLog2*2
WorkgroupLog2 if WorkgroupLog2 <= SubgroupLog2*2+2
SubgroupLog2*3 if WorkgroupLog2 >    SubgroupLog2*2+2
```

Reply:
Why `WorkgroupSizeLog2<=_SubgroupSizeLog2+4`? Left it out for now because it won't work with workgroup size 1024 and subgroup size 32

Reply:
corrected in my godbolt snippet